File: dart/lib/src/provider_service.dart
import 'dart:convert';

import 'package:trinsic_dart/src/proto/sdk/options/v1/options.pb.dart';
import 'package:trinsic_dart/src/proto/services/provider/v1/provider.pbgrpc.dart';
import 'package:trinsic_dart/src/service_base.dart';
import 'package:trinsic_dart/src/storage/token_provider.dart';

class ProviderService extends ServiceBase {
  late ProviderClient client;

  ProviderService(ServiceOptions? serverOptions, ITokenProvider? tokenProvider)
      : super(serverOptions, tokenProvider) {
    client = ProviderClient(super.channel);
  }

  Future<CreateEcosystemResponse> createEcosystem(
      {CreateEcosystemRequest? request}) async {
    request = request ?? CreateEcosystemRequest();
    var authenticate =
        (request.name.isNotEmpty || request.details.email.isNotEmpty);
    var metadata =
        authenticate ? buildMetadata(request: request) : buildMetadata();
    var response =
        await client.createEcosystem(request, options: await metadata);
    if (!response.profile.protection.enabled) {
      tokenProvider.Save(
          Base64Encoder.urlSafe().convert(response.profile.writeToBuffer()));
    }
    return response;
  }

// BEGIN Code generated by protoc-gen-trinsic. DO NOT EDIT.
// target: /home/runner/work/sdk/sdk/dart/lib/src/provider_service.dart

  Future<UpdateEcosystemResponse> updateEcosystem(
      UpdateEcosystemRequest request) async {
    ///  Update an existing ecosystem

    return client.updateEcosystem(request,
        options: await buildMetadata(request: request));
  }

  Future<GrantAuthorizationResponse> grantAuthorization(
      GrantAuthorizationRequest request) async {
    ///  Grant user authorization to ecosystem resources

    return client.grantAuthorization(request,
        options: await buildMetadata(request: request));
  }

  Future<RevokeAuthorizationResponse> revokeAuthorization(
      RevokeAuthorizationRequest request) async {
    ///  Revoke user authorization to ecosystem resources

    return client.revokeAuthorization(request,
        options: await buildMetadata(request: request));
  }

  Future<GetAuthorizationsResponse> getAuthorizations(
      GetAuthorizationsRequest request) async {
    ///  Retrieve the list of permissions for this particular account/ecosystem

    return client.getAuthorizations(request,
        options: await buildMetadata(request: request));
  }

  Future<AddWebhookResponse> addWebhook(AddWebhookRequest request) async {
    ///  Add a webhook endpoint to the ecosystem

    return client.addWebhook(request,
        options: await buildMetadata(request: request));
  }

  Future<DeleteWebhookResponse> deleteWebhook(
      DeleteWebhookRequest request) async {
    ///  Delete a webhook endpoint from the ecosystem

    return client.deleteWebhook(request,
        options: await buildMetadata(request: request));
  }

  Future<EcosystemInfoResponse> ecosystemInfo(
      EcosystemInfoRequest request) async {
    ///  Get ecosystem information

    return client.ecosystemInfo(request,
        options: await buildMetadata(request: request));
  }

  Future<GetPublicEcosystemInfoResponse> getPublicEcosystemInfo(
      GetPublicEcosystemInfoRequest request) async {
    ///  Get public ecosystem information about *any* ecosystem

    return client.getPublicEcosystemInfo(request,
        options: await buildMetadata());
  }

  Future<GenerateTokenResponse> generateToken(
      GenerateTokenRequest request) async {
    ///  Generates an unprotected authentication token that can be used to
    /// configure server side applications

    return client.generateToken(request,
        options: await buildMetadata(request: request));
  }

  Future<InviteResponse> invite(InviteRequest request) async {
    ///  Invite a user to the ecosystem

    return client.invite(request,
        options: await buildMetadata(request: request));
  }

  Future<InvitationStatusResponse> invitationStatus(
      InvitationStatusRequest request) async {
    ///  Check the status of an invitation

    return client.invitationStatus(request,
        options: await buildMetadata(request: request));
  }

  Future<GetOberonKeyResponse> getOberonKey(GetOberonKeyRequest request) async {
    ///  Returns the public key being used to create/verify oberon tokens

    return client.getOberonKey(request, options: await buildMetadata());
  }

  Future<GetEventTokenResponse> getEventToken(
      GetEventTokenRequest request) async {
    ///  Generate a signed token (JWT) that can be used to connect to the message bus

    return client.getEventToken(request,
        options: await buildMetadata(request: request));
  }

  Future<UpgradeDidResponse> upgradeDID(UpgradeDidRequest request) async {
    ///  Upgrade a wallet's DID from `did:key` to another method

    return client.upgradeDID(request,
        options: await buildMetadata(request: request));
  }

  Future<RetrieveDomainVerificationRecordResponse>
      retrieveDomainVerificationRecord() async {
    ///  Retrieve a random hash TXT that can be used to verify domain ownership
    var request = RetrieveDomainVerificationRecordRequest();
    return client.retrieveDomainVerificationRecord(request,
        options: await buildMetadata(request: request));
  }

  Future<RefreshDomainVerificationStatusResponse>
      refreshDomainVerificationStatus(
          RefreshDomainVerificationStatusRequest request) async {
    ///  Call to verify domain

    return client.refreshDomainVerificationStatus(request,
        options: await buildMetadata(request: request));
  }

  Future<SearchWalletConfigurationResponse> searchWalletConfigurations(
      SearchWalletConfigurationsRequest request) async {
    ///  Search for issuers/providers/verifiers in the current ecosystem

    return client.searchWalletConfigurations(request,
        options: await buildMetadata(request: request));
  }
// END Code generated by protoc-gen-trinsic. DO NOT EDIT.
}

File: web/src/ProviderService.ts
import ServiceBase from "./ServiceBase";
import * as proto from "./proto";

import type { Client as BrowserClient } from "nice-grpc-web";
import { Base64 } from "js-base64";

export class ProviderService extends ServiceBase {
    client: BrowserClient<typeof proto.ProviderDefinition>;

    constructor(options?: proto.ServiceOptions) {
        super(options);

        this.client = this.createClient(proto.ProviderDefinition);
    }

    public async createEcosystem(
        request: proto.CreateEcosystemRequest
    ): Promise<proto.CreateEcosystemResponse> {
        const response =
            request.name?.trim() || request.details?.email?.trim()
                ? await this.client.createEcosystem(request, {
                      metadata: await this.buildMetadata(
                          proto.CreateEcosystemRequest.encode(request).finish()
                      ),
                  })
                : await this.client.createEcosystem(request);

        const authToken = Base64.fromUint8Array(
            proto.AccountProfile.encode(response.profile!).finish(),
            true
        );
        this.options.authToken = authToken;
        return response;
    }
    // BEGIN Code generated by protoc-gen-trinsic. DO NOT EDIT.
    // target: /home/runner/work/sdk/sdk/web/src/ProviderService.ts

    /** Update an existing ecosystem */
    public async updateEcosystem(
        request: proto.UpdateEcosystemRequest
    ): Promise<proto.UpdateEcosystemResponse> {
        return this.client.updateEcosystem(request, {
            metadata: await this.buildMetadata(
                proto.UpdateEcosystemRequest.encode(request).finish()
            ),
        });
    }
    /** Grant user authorization to ecosystem resources */
    public async grantAuthorization(
        request: proto.GrantAuthorizationRequest
    ): Promise<proto.GrantAuthorizationResponse> {
        return this.client.grantAuthorization(request, {
            metadata: await this.buildMetadata(
                proto.GrantAuthorizationRequest.encode(request).finish()
            ),
        });
    }
    /** Revoke user authorization to ecosystem resources */
    public async revokeAuthorization(
        request: proto.RevokeAuthorizationRequest
    ): Promise<proto.RevokeAuthorizationResponse> {
        return this.client.revokeAuthorization(request, {
            metadata: await this.buildMetadata(
                proto.RevokeAuthorizationRequest.encode(request).finish()
            ),
        });
    }
    /** Retrieve the list of permissions for this particular account/ecosystem */
    public async getAuthorizations(
        request: proto.GetAuthorizationsRequest
    ): Promise<proto.GetAuthorizationsResponse> {
        return this.client.getAuthorizations(request, {
            metadata: await this.buildMetadata(
                proto.GetAuthorizationsRequest.encode(request).finish()
            ),
        });
    }
    /** Add a webhook endpoint to the ecosystem */
    public async addWebhook(
        request: proto.AddWebhookRequest
    ): Promise<proto.AddWebhookResponse> {
        return this.client.addWebhook(request, {
            metadata: await this.buildMetadata(
                proto.AddWebhookRequest.encode(request).finish()
            ),
        });
    }
    /** Delete a webhook endpoint from the ecosystem */
    public async deleteWebhook(
        request: proto.DeleteWebhookRequest
    ): Promise<proto.DeleteWebhookResponse> {
        return this.client.deleteWebhook(request, {
            metadata: await this.buildMetadata(
                proto.DeleteWebhookRequest.encode(request).finish()
            ),
        });
    }
    /** Get ecosystem information */
    public async ecosystemInfo(
        request: proto.EcosystemInfoRequest
    ): Promise<proto.EcosystemInfoResponse> {
        return this.client.ecosystemInfo(request, {
            metadata: await this.buildMetadata(
                proto.EcosystemInfoRequest.encode(request).finish()
            ),
        });
    }
    /** Get public ecosystem information about *any* ecosystem */
    public async getPublicEcosystemInfo(
        request: proto.GetPublicEcosystemInfoRequest
    ): Promise<proto.GetPublicEcosystemInfoResponse> {
        return this.client.getPublicEcosystemInfo(request, {
            metadata: await this.buildMetadata(),
        });
    }
    /** Generates an unprotected authentication token that can be used to
     * configure server side applications */
    public async generateToken(
        request: proto.GenerateTokenRequest
    ): Promise<proto.GenerateTokenResponse> {
        return this.client.generateToken(request, {
            metadata: await this.buildMetadata(
                proto.GenerateTokenRequest.encode(request).finish()
            ),
        });
    }
    /** Invite a user to the ecosystem */
    public async invite(
        request: proto.InviteRequest
    ): Promise<proto.InviteResponse> {
        return this.client.invite(request, {
            metadata: await this.buildMetadata(
                proto.InviteRequest.encode(request).finish()
            ),
        });
    }
    /** Check the status of an invitation */
    public async invitationStatus(
        request: proto.InvitationStatusRequest
    ): Promise<proto.InvitationStatusResponse> {
        return this.client.invitationStatus(request, {
            metadata: await this.buildMetadata(
                proto.InvitationStatusRequest.encode(request).finish()
            ),
        });
    }
    /** Returns the public key being used to create/verify oberon tokens */
    public async getOberonKey(
        request: proto.GetOberonKeyRequest
    ): Promise<proto.GetOberonKeyResponse> {
        return this.client.getOberonKey(request, {
            metadata: await this.buildMetadata(),
        });
    }
    /** Generate a signed token (JWT) that can be used to connect to the message bus */
    public async getEventToken(
        request: proto.GetEventTokenRequest
    ): Promise<proto.GetEventTokenResponse> {
        return this.client.getEventToken(request, {
            metadata: await this.buildMetadata(
                proto.GetEventTokenRequest.encode(request).finish()
            ),
        });
    }
    /** Upgrade a wallet's DID from `did:key` to another method */
    public async upgradeDID(
        request: proto.UpgradeDidRequest
    ): Promise<proto.UpgradeDidResponse> {
        return this.client.upgradeDID(request, {
            metadata: await this.buildMetadata(
                proto.UpgradeDidRequest.encode(request).finish()
            ),
        });
    }
    /** Retrieve a random hash TXT that can be used to verify domain ownership */
    public async retrieveDomainVerificationRecord(): Promise<proto.RetrieveDomainVerificationRecordResponse> {
        let request = proto.RetrieveDomainVerificationRecordRequest.fromPartial(
            {}
        );
        return this.client.retrieveDomainVerificationRecord(request, {
            metadata: await this.buildMetadata(
                proto.RetrieveDomainVerificationRecordRequest.encode(
                    request
                ).finish()
            ),
        });
    }
    /** Call to verify domain */
    public async refreshDomainVerificationStatus(
        request: proto.RefreshDomainVerificationStatusRequest
    ): Promise<proto.RefreshDomainVerificationStatusResponse> {
        return this.client.refreshDomainVerificationStatus(request, {
            metadata: await this.buildMetadata(
                proto.RefreshDomainVerificationStatusRequest.encode(
                    request
                ).finish()
            ),
        });
    }
    /** Search for issuers/providers/verifiers in the current ecosystem */
    public async searchWalletConfigurations(
        request: proto.SearchWalletConfigurationsRequest
    ): Promise<proto.SearchWalletConfigurationResponse> {
        return this.client.searchWalletConfigurations(request, {
            metadata: await this.buildMetadata(
                proto.SearchWalletConfigurationsRequest.encode(request).finish()
            ),
        });
    }
    // END Code generated by protoc-gen-trinsic. DO NOT EDIT.
}
